viscy/representation/evaluation.py

```diff
@@ -0,0 +1,381 @@
+    """
+    Evaluate the clustering of the embeddings compared to the ground truth labels.
+
+    Parameters
+    ----------
+    method : str, optional
+        Metric to use for evaluation ('nmi' or 'ari'). Default is 'nmi'.
+
+    Returns
+    -------
+    float
+        NMI or ARI score depending on the method chosen.
+    """
+    clusters = dbscan_clustering(embeddings)
+
+    if method == "nmi":
+        score = normalized_mutual_info_score(annotations, clusters)
+    elif method == "ari":
+        score = adjusted_rand_score(annotations, clusters)
+    else:
+        raise ValueError("Invalid method. Choose 'nmi' or 'ari'.")
+
+    return score
+
+
+def compute_pca(embedding_dataset, n_components=None, normalize_features=True):
+    features = embedding_dataset["features"]
+    projections = embedding_dataset["projections"]
+
+    if normalize_features:
+        scaled_projections = StandardScaler().fit_transform(projections.values)
+        scaled_features = StandardScaler().fit_transform(features.values)
+    else:
+        scaled_projections = projections.values
+        scaled_features = features.values
+
+    # Compute PCA with specified number of components
+    PCA_features = PCA(n_components=n_components, random_state=42)
+    PCA_projection = PCA(n_components=n_components, random_state=42)
+    pc_features = PCA_features.fit_transform(scaled_features)
+    pc_projection = PCA_projection.fit_transform(scaled_projections)
+
+    # Prepare DataFrame with id and PCA coordinates
+    pca_df = pd.DataFrame(
+        {
+            "id": embedding_dataset["id"].values,
+            "fov_name": embedding_dataset["fov_name"].values,
+            "PCA1": pc_features[:, 0],
+            "PCA2": pc_features[:, 1],
+            "PCA3": pc_features[:, 2],
+            "PCA4": pc_features[:, 3],
+            "PCA5": pc_features[:, 4],
+            "PCA6": pc_features[:, 5],
+            "PCA1_proj": pc_projection[:, 0],
+            "PCA2_proj": pc_projection[:, 1],
+            "PCA3_proj": pc_projection[:, 2],
+            "PCA4_proj": pc_projection[:, 3],
+            "PCA5_proj": pc_projection[:, 4],
+            "PCA6_proj": pc_projection[:, 5],
+        }
+    )
+
+    return PCA_features, PCA_projection, pca_df
+
+
+def compute_umap(embedding_dataset, normalize_features=True):
+    features = embedding_dataset["features"]
+    projections = embedding_dataset["projections"]
+
+    if normalize_features:
+        scaled_projections = StandardScaler().fit_transform(projections.values)
+        scaled_features = StandardScaler().fit_transform(features.values)
+    else:
+        scaled_projections = projections.values
+        scaled_features = features.values
+
+    # Compute UMAP for features and projections
+    # Computing 3 components to enable 3D visualization.
+    umap_features = umap.UMAP(random_state=42, n_components=3)
+    umap_projection = umap.UMAP(random_state=42, n_components=3)
```

Comment:
Why is `n_components` hard-coded to be 3 when we only plot in 2D?

Reply:
cc @Soorya19Pradeep @alishbaimran

Reply:
If we plot 2D UMAP they will look a bit different:
`n_components=2`:
![image](https://github.com/user-attachments/assets/bf94ef84-35db-4fb3-b048-35171212f423)
`n_components=3`:
![image](https://github.com/user-attachments/assets/7e4cc760-89d1-4c77-9989-66c13bf45d36)

Reply:
@ziw-liu , I have been using 2 UMAP components for plotting and 4 components when I compute correlation with computed features. Does this mean it will change the interpretation of results?

Reply:
@Soorya19Pradeep The 2- and 4-component UMAP values will be different.

Reply:
@ziw-liu , would you suggest using 4 or 2-component UMAP? Based on the observation you made on the plot I am guessing 2-component UMAP?

Reply:
I'm not sure how to interpret the correlation coefficient between UMAP and image features, can you provide an example in literature? I would use 2 components for visualization though.

Reply:
Moving this discussion to #165.